File: python/statistics.py
# statistics.py

def compute_gpa(grades):
    """Compute GPA: sum of grades divided by the number of subjects."""
    if not grades:
        return 0
    return sum(grades) / len(grades)

def generate_statistics(records):
    """Generate and print summary statistics for all student records, including grade distribution."""
    if not records:
        print("No records found!")
        return

    stats_list = []
    grade_counts = {'A': 0, 'B': 0, 'C': 0, 'D': 0, 'F': 0}
    
    for student_id, rec in records.items():
        gpa = compute_gpa(rec["grades"])
        stats_list.append((student_id, rec["name"], gpa))
        
        # Categorize GPA into grades
        if gpa >= 90:
            grade_counts['A'] += 1
        elif gpa >= 80:
            grade_counts['B'] += 1
        elif gpa >= 70:
            grade_counts['C'] += 1
        elif gpa >= 60:
            grade_counts['D'] += 1
        else:
            grade_counts['F'] += 1

    # Calculate average, highest, and lowest GPA
    total_gpa = sum(gpa for _, _, gpa in stats_list)
    average_gpa = total_gpa / len(stats_list) if stats_list else 0
    highest = max(stats_list, key=lambda x: x[2], default=None)
    lowest = min(stats_list, key=lambda x: x[2], default=None)

    print("\n=== Class Statistics ===")
    print(f"Average GPA: {average_gpa:.2f}")
    if highest:
        print(f"Highest GPA: {highest[2]:.2f} (Student: {highest[1]}, ID: {highest[0]})")
    if lowest:
        print(f"Lowest GPA: {lowest[2]:.2f} (Student: {lowest[1]}, ID: {lowest[0]})")
    
    # Print grade distribution
    print("\n=== Grade Distribution ===")
    for grade, count in grade_counts.items():
        print(f"{grade}: {count} student(s)")

def sort_students(records, sort_by='gpa'):
    """Sort students by GPA (descending) or name (ascending). Returns a sorted list."""
    stats_list = []
    for student_id, rec in records.items():
        gpa = compute_gpa(rec["grades"])
        stats_list.append((student_id, rec["name"], gpa))
    
    if sort_by == 'name':
        return sorted(stats_list, key=lambda x: (x[1].lower(), x[0]))
    else:
        return sorted(stats_list, key=lambda x: (-x[2], x[0]))

File: python/kotlin-app/app/src/main/kotlin/Statistics.kt
object Statistics {
    fun computeGPA(grades: List<Double>): Double {
        return if (grades.isEmpty()) 0.0 else grades.average()
    }

    fun generateStatistics(records: Map<String, StudentRecord>) {
        if (records.isEmpty()) {
            println("No records found!")
            return
        }

        val stats = records.map { (id, rec) ->
            Triple(id, rec.name, computeGPA(rec.grades))
        }

        val averageGPA = stats.map { it.third }.average()
        val highest = stats.maxByOrNull { it.third }
        val lowest = stats.minByOrNull { it.third }

        val gradeCounts = mutableMapOf(
            'A' to 0, 'B' to 0, 'C' to 0, 'D' to 0, 'F' to 0
        )

        stats.forEach { (_, _, gpa) ->
            when {
                gpa >= 90 -> gradeCounts['A'] = gradeCounts['A']!! + 1
                gpa >= 80 -> gradeCounts['B'] = gradeCounts['B']!! + 1
                gpa >= 70 -> gradeCounts['C'] = gradeCounts['C']!! + 1
                gpa >= 60 -> gradeCounts['D'] = gradeCounts['D']!! + 1
                else -> gradeCounts['F'] = gradeCounts['F']!! + 1
            }
        }

        println("\n=== Class Statistics ===")
        println("Average GPA: ${"%.2f".format(averageGPA)}")
        highest?.let {
            println("Highest GPA: ${"%.2f".format(it.third)} (Student: ${it.second}, ID: ${it.first})")
        }
        lowest?.let {
            println("Lowest GPA: ${"%.2f".format(it.third)} (Student: ${it.second}, ID: ${it.first})")
        }

        println("\n=== Grade Distribution ===")
        gradeCounts.forEach { (grade, count) ->
            println("$grade: $count student(s)")
        }
    }

    fun sortStudents(records: Map<String, StudentRecord>, sortBy: String): List<Triple<String, String, Double>> {
        val stats = records.map { (id, rec) ->
            Triple(id, rec.name, computeGPA(rec.grades))
        }

        return when (sortBy) {
            "name" -> stats.sortedWith(compareBy({ it.second.lowercase() }, { it.first }))
            else -> stats.sortedWith(compareByDescending<Triple<String, String, Double>> { it.third }.thenBy { it.first })
        }
    }
}